packages/shared/src/contracts.ts

```diff
@@ -53,22 +29,31 @@ export const getStateChainGatewayContractAddress = (
   return ADDRESSES[network].STATE_CHAIN_MANAGER_CONTRACT_ADDRESS;
 };
 
-export const requestApproval = async (
-  erc20Address: string,
-  approvalAddress: string,
+export const checkAllowance = async (
   amount: BigNumberish,
+  spenderAddress: string,
+  erc20Address: string,
   signer: Signer,
-): Promise<ContractReceipt | null> => {
+) => {
   const erc20 = ERC20__factory.connect(erc20Address, signer);
   const signerAddress = await signer.getAddress();
-  const allowance = await erc20.allowance(signerAddress, approvalAddress);
-
-  if (allowance.lt(amount)) {
-    const tx = await erc20.approve(approvalAddress, amount);
-    return tx.wait(1);
-  }
+  const allowance = await erc20.allowance(signerAddress, spenderAddress);
+  return { allowance, isAllowable: allowance.gte(amount), erc20 };
+};
 
-  return null;
+export const approve = async (
+  amount: BigNumberish,
+  spenderAddress: string,
+  erc20: ERC20,
+  allowance: BigNumberish,
+  nonce?: bigint | number | string,
+): Promise<ContractReceipt | null> => {
+  const amountBigNumber = BigNumber.from(amount);
+  const allowanceBigNumber = BigNumber.from(allowance);
+  if (allowanceBigNumber.gte(amountBigNumber)) return null;
+  const requiredAmount = amountBigNumber.sub(allowanceBigNumber);
+  const tx = await erc20.approve(spenderAddress, requiredAmount, { nonce });
+  return tx.wait(1);
```

Comment:
think we should try to handle transactions consistent - either we should always wait for inclusion before returning or return immediately after submitting

Reply:
I don't know anywhere we don't wait for inclusion before returning a hash to the consumer